README.md
# StatsD HTTP Proxy

StatsD HTTP proxy with REST interface for using in browsers

[![Go Report Card](https://goreportcard.com/badge/github.com/GoMetric/statsd-http-proxy?2)](https://goreportcard.com/report/github.com/GoMetric/statsd-http-proxy)
[![Build Status](https://travis-ci.org/GoMetric/statsd-http-proxy.svg?branch=master)](https://travis-ci.org/GoMetric/statsd-http-proxy)
[![Code Climate](https://codeclimate.com/github/GoMetric/statsd-http-proxy/badges/gpa.svg?1)](https://codeclimate.com/github/GoMetric/statsd-http-proxy)
[![Coverage Status](https://coveralls.io/repos/github/GoMetric/statsd-http-proxy/badge.svg?branch=master)](https://coveralls.io/github/GoMetric/statsd-http-proxy?branch=master)

StatsD uses UDP connections, and  can not be used directly from browser. This server is a HTTP proxy to StatsD, useful for sending metrics to StatsD from frontend by AJAX.

Requests may be optionally authenticated using JWT tokens.

## Table of contents

* [Installation](#installation)
* [Requirements](#requirements)
* [Proxy client for browser](#proxy-client-for-browser)
* [Nginx config](#nginx-config)
* [Usage](#usage)
* [Authentication](#authentication)
* [Rest resources](#rest-resources)
  * [Heartbeat](#heartbeat)
  * [Count](#count)
  * [Gauge](#gauge)
  * [Timing](#timing)
  * [Set](#set)
* [Response](#response)
* [Testing](#testing)
* [Benchmark](#benchmark)
* [Useful resources](#useful-resources)


## Installation

### Build from sources

```
git clone git@github.com:GoMetric/statsd-http-proxy.git
make build
```

### Build docker image

Build your own docker image: https://github.com/GoMetric/statsd-http-proxy-docker

### Official docker image

Use [Docker image](https://hub.docker.com/r/gometric/statsd-http-proxy/):

[![docker](https://img.shields.io/docker/pulls/gometric/statsd-http-proxy.svg?style=flat)](https://hub.docker.com/r/gometric/statsd-http-proxy/)

```
docker run -p 80:80 gometric/statsd-http-proxy:latest --verbose
```

Secure connection:

```
docker run -p 4433:4433 -v "$(pwd)":/certs/  gometric/statsd-http-proxy:latest --verbose --http-port=4433 --tls-cert=/certs/cert.pem --tls-key=/certs/key.pem
```

## Requirements

* [GoMetric/go-statsd-client](https://github.com/GoMetric/go-statsd-client) - StatsD client library for Go
* [dgrijalva/jwt-go](https://github.com/dgrijalva/jwt-go) - JSON Web Tokens builder and parser
* [gorilla/mux](https://github.com/gorilla/mux) - URL router and dispatcher

## Proxy client for browser

Basic implementation of proxy client may be found at https://github.com/GoMetric/statsd-http-proxy-client.

## Nginx config

Configuration of Nginx balancer:

```
server {
    listen 443 http2;

    server_name statsd-proxy.example.com;

    ssl on;
    ssl_certificate     /etc/pki/nginx/ssl.crt;
    ssl_certificate_key /etc/pki/nginx/ssl.key;

    upstream statsd_proxy {
        keepalive 100;
        server statsd-proxy-1:8825 max_fails=0;
        server statsd-proxy-2:8825 max_fails=0;
    }
    
    location / {
        proxy_pass http://statsd_proxy;
        proxy_redirect off;
        proxy_http_version 1.1;
        proxy_set_header Connection "keep-alive";
    }
}
```


## Usage

* Run server (HTTP):

```bash
statsd-http-proxy \
    --verbose \
    --http-host=127.0.0.1 \
    --http-port=8080 \
    --statsd-host=127.0.0.1 \
    --statsd-port=8125 \
    --jwt-secret=somesecret \
    --metric-prefix=prefix.subprefix
```

* Run server (HTTPS):

```bash
statsd-http-proxy \
    --verbose \
    --http-host=127.0.0.1 \
    --http-port=433 \
    --tls-cert=cert.pem \
    --tls-key=key.pem \
    --statsd-host=127.0.0.1 \
    --statsd-port=8125 \
    --jwt-secret=somesecret \
    --metric-prefix=prefix.subprefix
```

Print server version and exit:

```bash
statsd-http-proxy --version
```

Command line arguments:

| Parameter       | Description                          | Default value                                                                     |
|-----------------|--------------------------------------|-----------------------------------------------------------------------------------|
| verbose         | Print debug info to stderr           | Optional. Default false                                                           |
| http-host       | Host of HTTP server                  | Optional. Default 127.0.0.1. To accept connections on any interface, set to ""    |
| http-port       | Port of HTTP server                  | Optional. Default 80                                                              |
| http-timeout-read | The maximum duration in seconds for reading the entire request, including the body | Optional. Defaults to 1 second |
| http-timeout-write | The maximum duration in seconds before timing out writes of the respons | Optional. Defaults to 1 second  |
| http-timeout-idle | The maximum amount of time in seconds to wait for the next request when keep-alives are enabled | Optional. Defaults to 1 second |
| tls-cert        | TLS certificate for the HTTPS        | Optional. Default "" to use HTTP. If both tls-cert and tls-key set, HTTPS is used |
| tls-key         | TLS private key for the HTTPS        | Optional. Default "" to use HTTP. If both tls-cert and tls-key set, HTTPS is used |
| statsd-host     | Host of StatsD instance              | Optional. Default 127.0.0.1                                                       |
| statsd-port     | Port of StatsD instance              | Optional. Default 8125                                                            |
| jwt-secret      | JWT token secret                     | Optional. If not set, server accepts all connections                              |
| metric-prefix   | Prefix, added to any metric name     | Optional. If not set, do not add prefix                                           |
| version         | Print version of server and exit     | Optional                                                                          |

Sample code to send metric in browser with JWT token in header:

```javascript
$.ajax({
    url: 'http://127.0.0.1:8080/count/some.key.name',
    method: 'POST',
    headers: {
        'X-JWT-Token': 'some-jwt-token'
    },
    data: {
        value: 100500
    }
});
```

## Authentication

Authentication is optional. It based on passing JWT token to server, encrypted with secret, specified in `jwt-secret`
command line argument. If secret not configured in `jwt-secret`, then requests to server accepted without authentication.
Token sends to server in `X-JWT-Token` header or in `token` query parameter.

We recommend to use JWT tokens to prevent flood requests: you need to setup JWT token expiration time, and update JWT token in browser each time you get 403 in response.

## Rest resources

See [statsd ductmentation](https://github.com/statsd/statsd/blob/master/docs/metric_types.md) about supported types.

### Heartbeat
```
GET /heartbeat
```
If server working, it responds with `OK`

### Count
```
POST /count/{key}
X-JWT-Token: {tokenString}
value=1&sampleRate=1
```

| Parameter  | Description                          | Default value                      |
|------------|--------------------------------------|------------------------------------|
| value      | Value. Negative to decrease          | Optional. Default 1                |
| sampleRate | Sample rate to skip metrics          | Optional. Default to 1: accept all |

### Gauge

Gauge is an arbitrary value. 
Only the last value during a flush interval is flushed to the backend. If the gauge is not updated at the next flush, it will send the previous value.
Gauge also may be set relatively to previously stored value. 
Is `shift` not passed, then `value` used. 
If `value` not passed, used default value equals 1.
To set a gauge to a negative number you need first set it to 0.

Absolute value:
```
POST /gauge/{key}
X-JWT-Token: {tokenString}
value=1
```

Shift of previous value:
```
POST /gauge/{key}
X-JWT-Token: {tokenString}
shift=-1
```

| Parameter  | Description                                     | Default value                      |
|------------|-------------------------------------------------|------------------------------------|
| value      | Integer value                                   | Optional. Default 1                |
| shift      | Signed int, relative to previously stored value | Optional                           |

### Timing
```
POST /timing/{key}
X-JWT-Token: {tokenString}
time=1234567&sampleRate=1
```

| Parameter  | Description                                   | Default value                      |
|------------|-----------------------------------------------|------------------------------------|
| time       | Time in milliseconds                          | Required                           |
| sampleRate | Float sample rate to skip metrics from 0 to 1 | Optional. Default to 1: accept all |

### Set
```
POST /set/{key}
X-JWT-Token: {tokenString}
value=1
```

| Parameter  | Description                          | Default value                      |
|------------|--------------------------------------|------------------------------------|
| value      | Integer value                        | Optional. Default 1                |

## Response

Server sends `200 OK` if send success, even StatsD server is down.

Other HTTP status codes:

| CODE             | Description                             |
|------------------|-----------------------------------------|
| 400 Bad Request  | Invalid parameters specified            |
| 401 Unauthorized | Token not sent                          |
| 403 Forbidden    | Token invalid/expired                   |
| 404 Not found    | Invalid url requested                   |
| 405 Wrong method | Request method not allowed for resource |

## Testing

It is useful for testing to start `netcat` UDP server,
listening for connections and watch incoming metrics.
To start server run:
```
nc -kluv localhost 8125
```
To send comamnd to statsd, run:

```
echo "counters" | nc localhost 8125
```

## Benchmark

Machine for benchmarking:

```
Intel(R) Core(TM) i5-2450M CPU @ 2.50GHz Dual Core / 8 GB RAM
```

Os:

```
Linux hp 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

Sysctl:

```
sudo sysctl net.ipv4.ip_local_port_range="15000 61000"
sudo sysctl net.ipv4.tcp_fin_timeout=30
```

Benchmarked by siege v. 4.0.4

### Proxy CLI arguments

Without JWT token:

```
$ GOMAXPROCS=1 ./bin/statsd-http-proxy --http-host=127.0.0.1 --http-port=8080 --statsd-host=127.0.0.1 --statsd-port=8125
```

With JWT token:

```
$ GOMAXPROCS=1 ./bin/statsd-http-proxy --http-host=127.0.0.1 --http-port=8080 --statsd-host=127.0.0.1 --statsd-port=8125 --jwt-secret=somesecret
```

### Requests to proxy

#### Router: Gorilla MUX, Keep alive: disabled, JWT auth: disabled

```
siege -R <(echo connection = close) -c 255 -r 2000 "http://127.0.0.1:8080/count/a.b.c.d POST value=42"
```

#### Router: Gorilla MUX, Keep alive: enabled, JWT auth: disabled

```
$ siege -R <(echo connection = keep-alive) -c 255 -r 2000 "http://127.0.0.1:8080/count/a.b.c.d POST value=42"
```

#### Router: Gorilla MUX, Keep alive: disabled, JWT auth: enabled

```
$ siege -R <(echo connection = close) -c 255 -r 2000 -H 'X-JWT-Token:eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJzdGF0c2QtcmVzdC1zZXJ2ZXIiLCJpYXQiOjE1MDY5NzI1ODAsImV4cCI6MTg4NTY2Mzc4MCwiYXVkIjoiaHR0cHM6Ly9naXRodWIuY29tL3Nva2lsL3N0YXRzZC1yZXN0LXNlcnZlciIsInN1YiI6InNva2lsIn0.sOb0ccRBnN1u9IP2jhJrcNod14G5t-jMHNb_fsWov5c' "http://127.0.0.1:8080/count/a.b.c.d POST value=42"
```

#### Router: Gorilla MUX, Keep alive: enabled, JWT auth: enabled

```
$ siege -R <(echo connection = keep-alive) -c 255 -r 2000 -H 'X-JWT-Token:eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJzdGF0c2QtcmVzdC1zZXJ2ZXIiLCJpYXQiOjE1MDY5NzI1ODAsImV4cCI6MTg4NTY2Mzc4MCwiYXVkIjoiaHR0cHM6Ly9naXRodWIuY29tL3Nva2lsL3N0YXRzZC1yZXN0LXNlcnZlciIsInN1YiI6InNva2lsIn0.sOb0ccRBnN1u9IP2jhJrcNod14G5t-jMHNb_fsWov5c' "http://127.0.0.1:8080/count/a.b.c.d POST value=42"
```

### Results

Concurent 255 users made 2000 requests each. Total request count: 510000

| Router           | Keep alive | JWT      | Elapsed time | Transaction rate  | Concurrency |
|------------------|------------|----------|--------------|-------------------|-------------|
| GorillaMux 1.7.3 | disabled   | disabled | 94.73 secs   | 5383.72 trans/sec | 244.02      |
| GorillaMux 1.7.3 | enabled    | disabled | 55.70 secs   | 9156.19 trans/sec | 252.27      |
| GorillaMux 1.7.3 | disabled   | enabled  | 117.80 secs  | 4329.37 trans/sec | 245.98      |
| GorillaMux 1.7.3 | enabled    | enabled  | 77.97 secs   | 6540.98 trans/sec | 253.70      |
| HttpRouter 1.3.0 | disabled   | disabled | 92.93 secs   | 5487.99 trans/sec | 244.09      |
| HttpRouter 1.3.0 | enabled    | disabled | 54.87 secs   | 9294.70 trans/sec | 252.65      |
| HttpRouter 1.3.0 | disabled   | enabled  | 115.35 secs  | 4421.33 trans/sec | 245.48      |
| HttpRouter 1.3.0 | enabled    | enabled  | 75.14 secs   | 6787.33 trans/sec | 253.25      |


## Useful resources
* [https://github.com/etsy/statsd](https://github.com/etsy/statsd) - StatsD sources
* [Official Docker image for Graphite](https://github.com/graphite-project/docker-graphite-statsd)
* [Docker image with StatsD, Graphite, Grafana 2 and a Kamon Dashboard](https://github.com/kamon-io/docker-grafana-graphite)
* [Online JWT generator](http://jwtbuilder.jamiekurtz.com/)
* [Client for StatsD (Golang)](https://github.com/GoMetric/go-statsd-client)


CHANGELOG.md
## 0.12.1 (2021-04-27)
  * Use GoMetric StatsD by default

## 0.12.0 (2021-04-17)
  * Use HttpRouter by default

## 0.11.0 (2021-04-16)
  * Use Cactus StatsD Client

## 0.10.0 (2021-04-05)
  * Default HTTP port changed to 8825
  * Configure HTTP timeouts through `--http-timeout-read`, `--http-timeout-write` and `--http-timeout-idle`

## 0.9 (2019-01-29)
  * TLS Secure connection listening

## 0.8 (2018-02-24)
  * Added support of preflight CORS OPTIONS requests with header `X-JWT-Token`
  * Added support of authentication with passing token in query string instead of header `X-JWT-Token`

## 0.7 (2017-12-25)
  * Binary renamed to `statsd-http-proxy`


proxy/server_test.go
package proxy

import (
	"reflect"
	"testing"

	"github.com/stretchr/testify/require"
)

func TestNewServerWithGorillaMuxRouter(t *testing.T) {
	proxyServer := NewServer(
		"127.0.0.1",
		8080,
		1,
		1,
		1,
		"127.0.0.1",
		8125,
		"tls.crt",
		"tls.key",
		"someMetricPrefix",
		"someTokenSecret",
		true,
		"GorillaMux",
		"Cactus",
	)

	require.Equal(t, "*proxy.Server", reflect.TypeOf(proxyServer).String())

	require.Equal(t, "127.0.0.1:8080", proxyServer.httpAddress)

	require.Equal(t, "*statsdclient.CactusStatsdClientAdapter", reflect.TypeOf(proxyServer.statsdClient).String())

	require.Equal(t, "tls.crt", proxyServer.tlsCert)

	require.Equal(t, "tls.key", proxyServer.tlsKey)
}

func TestNewServerWithHttpRouter(t *testing.T) {
	proxyServer := NewServer(
		"127.0.0.1",
		8080,
		1,
		1,
		1,
		"127.0.0.1",
		8125,
		"tls.crt",
		"tls.key",
		"someMetricPrefix",
		"someTokenSecret",
		true,
		"HttpRouter",
		"Cactus",
	)

	require.Equal(t, "*proxy.Server", reflect.TypeOf(proxyServer).String())

	require.Equal(t, "127.0.0.1:8080", proxyServer.httpAddress)

	require.Equal(t, "*statsdclient.CactusStatsdClientAdapter", reflect.TypeOf(proxyServer.statsdClient).String())

	require.Equal(t, "tls.crt", proxyServer.tlsCert)

	require.Equal(t, "tls.key", proxyServer.tlsKey)
}


proxy/server.go
package proxy

import (
	"context"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/GoMetric/statsd-http-proxy/proxy/routehandler"
	"github.com/GoMetric/statsd-http-proxy/proxy/router"
	"github.com/GoMetric/statsd-http-proxy/proxy/statsdclient"
)

// Server is a proxy server between HTTP REST API and UDP Connection to StatsD
type Server struct {
	httpAddress  string
	httpServer   *http.Server
	statsdClient statsdclient.StatsdClientInterface
	tlsCert      string
	tlsKey       string
}

// NewServer creates new instance of StatsD HTTP Proxy
func NewServer(
	httpHost string,
	httpPort int,
	httpReadTimeout int,
	httpWriteTimeout int,
	httpIdleTimeout int,
	statsdHost string,
	statsdPort int,
	tlsCert string,
	tlsKey string,
	metricPrefix string,
	tokenSecret string,
	verbose bool,
	httpRouterName string,
	statsdClientName string,
) *Server {
	// configure logging
	var logOutput io.Writer
	if verbose == true {
		logOutput = os.Stderr
	} else {
		logOutput = ioutil.Discard
	}

	log.SetOutput(logOutput)

	logger := log.New(logOutput, "", log.LstdFlags)

	// create StatsD Client
	var statsdClient statsdclient.StatsdClientInterface
	switch statsdClientName {
	case "Cactus":
		statsdClient = statsdclient.NewCactusClient(statsdHost, statsdPort)
	case "GoMetric":
		statsdClient = statsdclient.NewGoMetricClient(statsdHost, statsdPort)
	default:
		panic("Passed statsd client not supported")
	}

	// build route handler
	routeHandler := routehandler.NewRouteHandler(
		statsdClient,
		metricPrefix,
	)

	// build router
	var httpServerHandler http.Handler
	switch httpRouterName {
	case "HttpRouter":
		httpServerHandler = router.NewHTTPRouter(routeHandler, tokenSecret)
	case "GorillaMux":
		httpServerHandler = router.NewGorillaMuxRouter(routeHandler, tokenSecret)

	default:
		panic("Passed HTTP router not supported")
	}

	// get HTTP server address to bind
	httpAddress := fmt.Sprintf("%s:%d", httpHost, httpPort)

	// create http server
	httpServer := &http.Server{
		Addr:           httpAddress,
		Handler:        httpServerHandler,
		ErrorLog:       logger,
		ReadTimeout:    time.Duration(httpReadTimeout) * time.Second,
		WriteTimeout:   time.Duration(httpWriteTimeout) * time.Second,
		IdleTimeout:    time.Duration(httpIdleTimeout) * time.Second,
		MaxHeaderBytes: 1 << 11,
	}

	statsdHTTPProxyServer := Server{
		httpAddress,
		httpServer,
		statsdClient,
		tlsCert,
		tlsKey,
	}

	return &statsdHTTPProxyServer
}

// Listen starts listening HTTP connections
func (proxyServer *Server) Listen() {
	// prepare for gracefull shutdown
	gracefullStopSignalHandler := make(chan os.Signal, 1)
	signal.Notify(gracefullStopSignalHandler, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)

	// start HTTP/HTTPS proxy to StatsD
	go func() {
		log.Printf("Starting HTTP server at %s", proxyServer.httpAddress)

		// open StatsD connection
		proxyServer.statsdClient.Open()
		defer proxyServer.statsdClient.Close()

		// open HTTP connection
		var err error
		if len(proxyServer.tlsCert) > 0 && len(proxyServer.tlsKey) > 0 {
			err = proxyServer.httpServer.ListenAndServeTLS(proxyServer.tlsCert, proxyServer.tlsKey)
		} else {
			err = proxyServer.httpServer.ListenAndServe()
		}

		if err != nil && err != http.ErrServerClosed {
			log.Fatal(err)
			log.Fatal("Can not start HTTP server")
		}
	}()

	<-gracefullStopSignalHandler

	// Gracefull shutdown
	log.Printf("Stopping HTTP server")

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer func() {
		cancel()
	}()

	if err := proxyServer.httpServer.Shutdown(ctx); err != nil {
		log.Fatalf("HTTP Server Shutdown Failed:%+v", err)
	}

	log.Printf("HTTP server stopped successfully")
}


proxy/statsdclient/cactus.go
package statsdclient

import (
	"strconv"
	"time"

	cactus "github.com/cactus/go-statsd-client/v5/statsd"
)

type CactusStatsdClientAdapter struct {
	cactusClient cactus.Statter
}

func (a *CactusStatsdClientAdapter) Open() {

}

func (a *CactusStatsdClientAdapter) Close() {
	a.cactusClient.Close()
}

func (a *CactusStatsdClientAdapter) Count(key string, value int, sampleRate float32) {
	a.cactusClient.Inc(key, int64(value), sampleRate)
}

func (a *CactusStatsdClientAdapter) Timing(key string, time int64, sampleRate float32) {
	a.cactusClient.Timing(key, time, sampleRate)
}

func (a *CactusStatsdClientAdapter) Gauge(key string, value int) {
	a.cactusClient.Gauge(key, int64(value), 1)
}

func (a *CactusStatsdClientAdapter) GaugeShift(key string, value int) {
	a.cactusClient.GaugeDelta(key, int64(value), 1)
}

func (a *CactusStatsdClientAdapter) Set(key string, value int) {
	a.cactusClient.SetInt(key, int64(value), 1)
}

func NewCactusClient(
	statsdHost string,
	statsdPort int,
) StatsdClientInterface {
	cactusClient, _ := cactus.NewBufferedClient(
		statsdHost+":"+strconv.Itoa(statsdPort),
		"",
		100*time.Microsecond,
		1432,
	)

	return &CactusStatsdClientAdapter{
		cactusClient: cactusClient,
	}
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	_ "net/http/pprof"
	"os"
	"runtime"

	"github.com/GoMetric/statsd-http-proxy/proxy"
)

// Version is a current git commit hash and tag
// Injected by compilation flag
var Version = "Unknown"

// BuildNumber is a current commit hash
// Injected by compilation flag
var BuildNumber = "Unknown"

// BuildDate is a date of build
// Injected by compilation flag
var BuildDate = "Unknown"

// HTTP connection params
const defaultHTTPHost = "127.0.0.1"
const defaultHTTPPort = 8825
const defaultHTTPReadTimeout = 1
const defaultHTTPWriteTimeout = 1
const defaultHTTPIdleTimeout = 1

// StatsD connection params
const defaultStatsDHost = "127.0.0.1"
const defaultStatsDPort = 8125

func main() {
	// declare command line options
	var httpHost = flag.String("http-host", defaultHTTPHost, "HTTP Host")
	var httpPort = flag.Int("http-port", defaultHTTPPort, "HTTP Port")
	var httpReadTimeout = flag.Int("http-timeout-read", defaultHTTPReadTimeout, "The maximum duration in seconds for reading the entire request, including the body")
	var httpWriteTimeout = flag.Int("http-timeout-write", defaultHTTPWriteTimeout, "The maximum duration in seconds before timing out writes of the respons")
	var httpIdleTimeout = flag.Int("http-timeout-idle", defaultHTTPIdleTimeout, "The maximum amount of time in seconds to wait for the next request when keep-alives are enabled")
	var tlsCert = flag.String("tls-cert", "", "TLS certificate to enable HTTPS")
	var tlsKey = flag.String("tls-key", "", "TLS private key  to enable HTTPS")
	var statsdHost = flag.String("statsd-host", defaultStatsDHost, "StatsD Host")
	var statsdPort = flag.Int("statsd-port", defaultStatsDPort, "StatsD Port")
	var metricPrefix = flag.String("metric-prefix", "", "Prefix of metric name")
	var tokenSecret = flag.String("jwt-secret", "", "Secret to encrypt JWT")
	var verbose = flag.Bool("verbose", false, "Verbose")
	var version = flag.Bool("version", false, "Show version")
	var httpRouterName = flag.String("http-router-name", "HttpRouter", "Type of HTTP router. Allowed values are GorillaMux and HttpRouter. Do not use in production.")
	var statsdClientName = flag.String("statsd-client-name", "GoMetric", "Type of StatsD client. Allowed values are Cactus and GoMetric. Do not use in production.")
	var profilerHTTPort = flag.Int("profiler-http-port", 0, "Start profiler localhost")

	// get flags
	flag.Parse()

	// show version and exit
	if *version {
		fmt.Printf(
			"StatsD HTTP Proxy v.%s, build %s from %s\n",
			Version,
			BuildNumber,
			BuildDate,
		)
		os.Exit(0)
	}

	// log build version
	log.Printf(
		"Starting StatsD HTTP Proxy v.%s, build %s from %s\n",
		Version,
		BuildNumber,
		BuildDate,
	)

	// start profiler
	if *profilerHTTPort > 0 {
		// enable block profiling
		runtime.SetBlockProfileRate(1)

		// start debug server
		profilerHTTPAddress := fmt.Sprintf("localhost:%d", *profilerHTTPort)
		go func() {
			log.Println("Profiler started at " + profilerHTTPAddress)
			log.Println("Open 'http://" + profilerHTTPAddress + "/debug/pprof/' in you browser or use 'go tool pprof http://" + profilerHTTPAddress + "/debug/pprof/heap' from console")
			log.Println("See details about pprof in https://golang.org/pkg/net/http/pprof/")
			log.Println(http.ListenAndServe(profilerHTTPAddress, nil))
		}()
	}

	// start proxy server
	proxyServer := proxy.NewServer(
		*httpHost,
		*httpPort,
		*httpReadTimeout,
		*httpWriteTimeout,
		*httpIdleTimeout,
		*statsdHost,
		*statsdPort,
		*tlsCert,
		*tlsKey,
		*metricPrefix,
		*tokenSecret,
		*verbose,
		*httpRouterName,
		*statsdClientName,
	)

	proxyServer.Listen()
}
